src/ConnectedMode/Persistence/ServerConnectionsRepository.cs

```diff
@@ -0,0 +1,207 @@
+            return wasAdded;
+        });
+    }
+
+    public bool TryDelete(string connectionId)
+    {
+        ServerConnection removedConnection = null;
+        var wasDeleted = SafeUpdateConnectionsFile(() => ServerConnectionsCache.TryRemove(connectionId, out removedConnection));
+        if (wasDeleted && removedConnection != null)
+        {
+            credentialsLoader.DeleteCredentials(removedConnection.CredentialsUri);
+        }
+        return wasDeleted;
+    }
+
+    public bool TryUpdateSettingsById(string connectionId, ServerConnectionSettings connectionSettings)
+    {
+        return SafeUpdateConnectionsFile(() => TryUpdateConnectionSettings(connectionId, connectionSettings));
+    }
+
+    public bool TryUpdateCredentialsById(string connectionId, ICredentials credentials)
+    {
+        try
+        {
+            var wasFound = TryGet(connectionId, out ServerConnection serverConnection);
+            if (!wasFound)
+            {
+                return false;
+            }
+            credentialsLoader.Save(credentials, serverConnection.CredentialsUri);
+            return true;
+        }
+        catch (Exception ex)
+        {
+            logger.WriteLine(ex.Message);
+            return false;
+        }
+    }
+
+    private bool TryUpdateConnectionSettings(string connectionId, ServerConnectionSettings connectionSettings)
+    {
+        var wasFound = ServerConnectionsCache.TryGetValue(connectionId, out ServerConnection existingServerConnection);
+        if (!wasFound)
+        {
+            return false;
+        }
+
+        existingServerConnection.Settings = connectionSettings;
+        return true;
+    }
+
+    private bool SafeUpdateConnectionsFile(Func<bool> updateConnectionsCache)
+    {
+        try
+        {
+            ReadServerConnectionsFile();
+            ServerConnectionsCache ??= new ConcurrentDictionary<string, ServerConnection>();
+
+            var wasUpdated = updateConnectionsCache();
+            if (!wasUpdated)
+            {
+                return false;
+            }
+
+            var model = serverConnectionModelMapper.GetServerConnectionsListJsonModel(ServerConnectionsCache.Values);
+
+            return jsonFileHandle.TryWriteToFile(storageFilePath, model);
+        }
+        catch (Exception ex)
+        {
+            logger.WriteLine(ex.Message);
+            return false;
+        }
+    }
+
+    private static string GetStorageFilePath(IEnvironmentVariableProvider environmentVariables)
+    {
+        var appDataFolder = environmentVariables.GetSLVSAppDataRootPath();
+        return Path.Combine(appDataFolder, ConnectionsFileName);
+    }
+
+    private void ReadServerConnectionsFile()
+    {
+        try
+        {
+            if (!jsonFileHandle.TryReadFile(storageFilePath, out ServerConnectionsListJsonModel model))
+            {
+                return;
+            }
+
+            ServerConnectionsCache = new ConcurrentDictionary<string, ServerConnection>();
```

Comment:
Multiple threads can assign this variable. Maybe Interlocked.CompareExchange would work here if you want to avoid locks, not sure. Overall it might be worth it to simplify the code by just using a normal Dictionary with a lock

Reply:
This is a good point for not having the dictionary null (your comment below regarding the null check). :)) 

But if I initialize the dictionary always, then the TryGet and GetAll will always read the file when no connections exist. Unless I introduce a flag, which I personally find ugly, but I guess is the only way.